### PR TITLE
gh-100562: improve performance of `pathlib.Path.absolute()`

### LINKED ISSUE #100562: Improve performance of pathlib.Path.absolute()
The current implementation of `pathlib.Path.absolute()` calls `self.cwd()` rather than `os.getcwd()`, and so constructs _two_ `Path` objects rather than one. As path objects are [slow to construct](https://github.com/faster-cpython/ideas/discussions/194), this has a performance impact.

### Linked PRs
* gh-100563

### PR DESCRIPTION
**Pull request by barneygale**
Increase performance of the `absolute()` method by calling `os.getcwd()` directly, rather than using the `Path.cwd()` class method. This avoids constructing an extra `Path` object (and the parsing/normalization that comes with it).

Decrease performance of the `cwd()` class method by calling the `Path.absolute()` method, rather than using `os.getcwd()` directly. This involves constructing an extra `Path` object. We do this to maintain a longstanding pattern where `os` functions are called from only one place, which allows them to be more readily replaced by users. As `cwd()` is generally called at most once within user programs, it's a good bargain.

```shell
# before
$ ./python -m timeit -s 'from pathlib import Path; p = Path("foo", "bar")' 'p.absolute()'
50000 loops, best of 5: 9.04 usec per loop
# after
$ ./python -m timeit -s 'from pathlib import Path; p = Path("foo", "bar")' 'p.absolute()'
50000 loops, best of 5: 5.02 usec per loop
```

* Issue: gh-100562

Automerge-Triggered-By: GH:AlexWaygood

### CODE REVIEW

File: Lib/pathlib.py
Comment (eryksun) [inline Lib/pathlib.py:830]: This overlooks relative drive paths, such as "Z:spam". If `self._drv[1:2] == ':'`, the drive working directory can be resolved with `nt._getfullpathname(self._drv)`. Otherwise use `os.getcwd()`.

Comment (barneygale) [reply Lib/pathlib.py:830]: If that's true, it's a bug in `ntpath.join()`, which we use to join arguments: https://github.com/python/cpython/blob/08e5594cf3d42391a48e0311f6b9393ec2e00e1e/Lib/pathlib.py#L293

Comment (barneygale) [reply Lib/pathlib.py:830]: Nevermind, I see what you mean. Could you log it as a separate bug? It's not a regression in this patch, right?

Comment (eryksun) [reply Lib/pathlib.py:830]: Okay. But the suggested fix should work in 3.12 now, right? This was discussed before on another issue, and it was blocked by a bug in `_from_parts()`. Fixing the bug in 3.10 and 3.11 will be more work.

Comment (barneygale) [reply Lib/pathlib.py:830]: Gotcha, I hadn't considered that. Thank you. If you log a bug I'll mark this PR as a draft and get that change made first.

File: Lib/pathlib.py
Comment (merwok) [inline Lib/pathlib.py:756]: Same remark as other PR about feeling weird about a straightforward implementation matching docstring being replaced by an indirect implementation.

Could you add a note in `absolute` to mark that `cwd` depends on `absolute` using `os.getcwd`?

Comment (barneygale) [reply Lib/pathlib.py:756]: I've added a comment in `cwd()` and adjusted its docstring. Strictly speaking `cwd()` _does not_ depend on `absolute()` using `os.getcwd()`. It relies on `absolute()` returning an absolute path per the POSIX definition (4.11 [here](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap04.html)). We may change the implementation of `absolute()` to use `nt._getfullpathname()` on Windows in future, but this would not break `cwd()`.

Comment (merwok) [reply Lib/pathlib.py:756]: The comment is great.  Thank you for all your pathlib efforts!

File: Misc/NEWS.d/next/Library/2022-12-28-00-28-43.gh-issue-100562.Hic0Z0.rst
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2022-12-28-00-28-43.gh-issue-100562.Hic0Z0.rst]: - State just how significant the performance improvement is
- Lay out more explicitly the trade-off here

```suggestion
Improve performance of :meth:`pathlib.Path.absolute` by nearly 2x. This comes
at the cost of a performance regression in :meth:`pathlib.Path.cwd`, which is
generally used less frequently in user code.
```

**Review by AlexWaygood — APPROVED** [review]
This looks great to me. I agree that it's much more important to optimise `Path.absolute()` than `Path.cwd()`, and I can reproduce the huge speedup.

Is the plan still to hold off merging this for now until https://github.com/python/cpython/pull/100563#discussion_r1058004995 is resolved, or can this be merged at any time?

**Comment by barneygale** [comment]
Thanks for the review!

> Is the plan still to hold off merging this for now until [#100563 (comment)](https://github.com/python/cpython/pull/100563#discussion_r1058004995) is resolved, or can this be merged at any time?

I think this can be merged any time. The backport for the fix for Eryk's issue will be slightly more complicated, but I will cope :D

**Comment by AlexWaygood** [comment]
I think all the tests are failing because Guido just broke the `main` branch accidentally :-)

But I'll merge this as soon as the fix in #100778 is merged, then we can re-run the tests here just to be on the safe side!

**Pull request merged**
